csharp/ql/src/semmle/code/csharp/dataflow/FrameworkDataFlow.qll

```diff
@@ -0,0 +1,313 @@
+    exists(Parameter p |
+      p = c.getParameter(i) and
+      if isCollectionType(p.getType())
+      then inputStack = ContentList::element()
+      else inputStack = ContentList::empty()
+    )
+  }
+
+  /**
+   * Gets an input specification that specifies output from the delegate at
+   * argument `i` as the input.
+   */
+  SummaryInput delegate(int i) { result = TDelegateSummaryInput(i) }
+
+  /**
+   * Gets an input specification that specifies output from the delegate at
+   * argument `i` as the input.
+   *
+   * `c` must be a compatible callable, that is, a callable where the `i`th
+   * parameter is a delegate.
+   */
+  SummaryInput delegate(FrameworkCallable c, int i) {
+    result = delegate(i) and
+    hasDelegateArgumentPosition(c, i)
+  }
+}
+
+class SummaryOutput = Impl::SummaryOutput;
+
+/** Provides predicates for constructing flow-summary output specifications. */
+module SummaryOutput {
+  /**
+   * Gets an output specification that specifies the qualifier in a call as
+   * the output.
+   */
+  SummaryOutput qualifier() { result = TQualifierSummaryOutput() }
+
+  /**
+   * Gets an output specification that specifies the return value from a call as
+   * the output.
+   */
+  SummaryOutput return() { result = TReturnSummaryOutput() }
+
+  /**
+   * Gets an output specification that specifies the `i`th argument in a call as
+   * the output.
+   */
+  SummaryOutput argument(int i) { result = TArgumentSummaryOutput(i) }
+
+  /**
+   * Gets an output specification that specifies parameter `j` of the delegate at
+   * argument `i` as the output.
+   */
+  SummaryOutput delegate(FrameworkCallable callable, int i, int j) {
+    result = TDelegateSummaryOutput(i, j) and
+    hasDelegateArgumentPosition2(callable, i, j)
+  }
+}
+
+class FlowSource = Impl::FlowSource;
+
+/** Provides predicates for constructing flow-source specifications. */
+module FlowSource {
+  class Range = Impl::FlowSource::Range;
+
+  private class RemoteFlowSource extends Range {
+    RemoteFlowSource() { this = "remote" }
+  }
+
+  FlowSource remote() { result = any(RemoteFlowSource s).toFlowSource() }
+
+  private class LocalFlowSource extends Range {
+    LocalFlowSource() { this = "local" }
+  }
+
+  FlowSource local() { result = any(LocalFlowSource s).toFlowSource() }
+
+  private class StoredFlowSource extends Range {
+    StoredFlowSource() { this = "stored" }
+  }
+
+  FlowSource stored() { result = any(StoredFlowSource s).toFlowSource() }
+}
+
+class FlowSink = Impl::FlowSink;
+
+/** Provides predicates for constructing flow-sink specifications. */
+module FlowSink {
+  class Range = Impl::FlowSink::Range;
+
+  private class RemoteFlowSink extends Range {
+    RemoteFlowSink() { this = "remote" }
+  }
+
+  FlowSink remote() { result = any(RemoteFlowSink s).toFlowSink() }
+
+  private class HtmlFlowSink extends Range {
+    HtmlFlowSink() { this = "html" }
+
+    override predicate isSubsetOf(FlowSink fs) { fs = remote() }
```

Comment:
Subsetting of classes usually goes via QL's class hierarchy mechanism. Is it necessary to build a custom system?

Reply:
I couldn't come up with a good way to do it. For example, I want to have flow sinks `Email`, `Html`, and `Remote`, where `Email` and `Html` are also `Remote` flow sinks, but not vice versa. Making `Remote` an abstract class does not work, because we want to be able to model remote flow sinks that are neither `Email` nor `Html`.

Reply:
I don't understand what you're saying from "not vice versa" and on. If we want a remote flow sink that's not `Email` or `Html` then we extend the abstract class `Remote`, and we have it.

Reply:
That would probably mean adding very specific flow sinks, or perhaps just a `NonOfTheOtherRemoteFlowSink`, but perhaps that is ok.